From 76983b3146c5c381c86e879a7a6cc114b41ef921 Mon Sep 17 00:00:00 2001
From: Jan Catrysse <jan.catrysse@geoxyz.eu>
Date: Thu, 3 Sep 2026 05:29:18 +0000
Subject: [PATCH] Target version filter offers the versions of the subprojects
 in the query (#43534).

A project issue list shows the issues of its subprojects by default, and
those issues can carry a target version belonging to a subproject and
shared with nobody. The list displays that version in its column, but the
target version filter does not offer it, because the value list is
project.shared_versions -- the versions the project can assign, not the
versions the query can return.

fixed_version_values now unions project.shared_versions with the visible
versions of the projects the query covers, so nothing that used to be
filterable stops being filterable: a version shared into the project from
outside the queried tree, such as a system-wide shared version, is kept.

The value list of a remote filter is fetched from QueriesController#filter
when the filter is added to a page that is already loaded. That action
built a query carrying only the project, so it could not know which
subprojects the query was scoped to; it now builds the query from the
request, after the permission check, and addFilter sends the filter form
along with the field name.
---
 app/assets/javascripts/application-legacy.js |  5 +-
 app/controllers/queries_controller.rb        |  2 +
 app/models/query.rb                          |  4 +-
 test/functional/queries_controller_test.rb   | 21 ++++++++
 test/unit/query_test.rb                      | 51 ++++++++++++++++++++
 5 files changed, 81 insertions(+), 2 deletions(-)

diff --git a/app/assets/javascripts/application-legacy.js b/app/assets/javascripts/application-legacy.js
index d8317b3f1..0ae6529e8 100644
--- a/app/assets/javascripts/application-legacy.js
+++ b/app/assets/javascripts/application-legacy.js
@@ -175,7 +175,10 @@ function addFilter(field, operator, values) {
   if (!filterOptions) return;
 
   if (filterOptions['remote'] && filterOptions['values'] == null) {
-    $.getJSON(filtersUrl, {'name': field}).done(function(data) {
+    var params = $('#filters-table').closest('form').serializeArray();
+    params.push({name: 'name', value: field});
+
+    $.getJSON(filtersUrl, params).done(function(data) {
       filterOptions['values'] = data;
       addFilter(field, operator, values) ;
     });
diff --git a/app/controllers/queries_controller.rb b/app/controllers/queries_controller.rb
index 24f37eda2..5c1adaa3e 100644
--- a/app/controllers/queries_controller.rb
+++ b/app/controllers/queries_controller.rb
@@ -100,6 +100,8 @@ class QueriesController < ApplicationController
       raise Unauthorized
     end
 
+    q.build_from_params(params)
+
     filter = q.available_filters[params[:name].to_s]
     values = filter ? filter.values : []
 
diff --git a/app/models/query.rb b/app/models/query.rb
index 7558f7409..a840f1ed2 100644
--- a/app/models/query.rb
+++ b/app/models/query.rb
@@ -660,7 +660,9 @@ class Query < ApplicationRecord
   def fixed_version_values
     versions = []
     if project
-      versions = project.shared_versions.to_a
+      versions =
+        project.shared_versions.to_a |
+        Version.visible.preload(:project).where(project_statement).to_a
     else
       versions = Version.visible.to_a
     end
diff --git a/test/functional/queries_controller_test.rb b/test/functional/queries_controller_test.rb
index 297bfa232..66a889881 100644
--- a/test/functional/queries_controller_test.rb
+++ b/test/functional/queries_controller_test.rb
@@ -823,6 +823,27 @@ class QueriesControllerTest < Redmine::ControllerTest
     assert_include ["eCookbook - 2.0", "3", "open"], json
   end
 
+  def test_filter_should_take_the_current_filters_into_account
+    version = Version.create!(:project => Project.find(3), :name => 'Unshared subproject version', :status => 'open')
+
+    @request.session[:user_id] = 2
+    get(
+      :filter,
+      :params => {
+        :project_id => 1,
+        :name => 'fixed_version_id',
+        :set_filter => '1',
+        :f => ['subproject_id'],
+        :op => {'subproject_id' => '='},
+        :v => {'subproject_id' => [version.project_id.to_s]}
+      }
+    )
+    assert_response :success
+    assert_equal 'application/json', response.media_type
+    json = ActiveSupport::JSON.decode(response.body)
+    assert_include ["eCookbook Subproject 1 - Unshared subproject version", version.id.to_s, "open"], json
+  end
+
   def test_version_filter_time_entries_with_project_id_should_return_filter_values
     @request.session[:user_id] = 2
     get(
diff --git a/test/unit/query_test.rb b/test/unit/query_test.rb
index ebf7fc991..68aec1307 100644
--- a/test/unit/query_test.rb
+++ b/test/unit/query_test.rb
@@ -161,6 +161,57 @@ class QueryTest < ActiveSupport::TestCase
     assert_include subproject_version.id.to_s, filter[:values].map(&:second)
   end
 
+  def test_fixed_version_filter_should_include_versions_shared_from_outside_the_project_tree
+    system_shared_version = Version.find(7)
+    query = IssueQuery.new(:project => Project.find(1), :name => '_')
+    filter = query.available_filters["fixed_version_id"]
+    assert_not_nil filter
+    assert_include system_shared_version.id.to_s, filter[:values].map(&:second)
+  end
+
+  def test_fixed_version_filter_should_include_subproject_versions_when_displaying_subproject_issues
+    with_settings :display_subprojects_issues => '1' do
+      version = Version.create!(:project => Project.find(3), :name => 'Unshared subproject version')
+      query = IssueQuery.new(:project => Project.find(1), :name => '_')
+      filter = query.available_filters["fixed_version_id"]
+      assert_not_nil filter
+      assert_include version.id.to_s, filter[:values].map(&:second)
+    end
+  end
+
+  def test_fixed_version_filter_should_not_include_subproject_versions_when_not_displaying_subproject_issues
+    with_settings :display_subprojects_issues => '0' do
+      version = Version.create!(:project => Project.find(3), :name => 'Unshared subproject version')
+      query = IssueQuery.new(:project => Project.find(1), :name => '_')
+      filter = query.available_filters["fixed_version_id"]
+      assert_not_nil filter
+      assert_not_include version.id.to_s, filter[:values].map(&:second)
+    end
+  end
+
+  def test_fixed_version_filter_should_respect_selected_subprojects
+    version1 = Version.create!(:project => Project.find(3), :name => 'Subproject 1 version')
+    version2 = Version.create!(:project => Project.find(4), :name => 'Subproject 2 version')
+    query = IssueQuery.new(:project => Project.find(1), :name => '_')
+    query.add_filter('subproject_id', '=', [Project.find(3).id.to_s])
+    filter = query.available_filters["fixed_version_id"]
+    assert_not_nil filter
+    values = filter[:values].map(&:second)
+    assert_include version1.id.to_s, values
+    assert_not_include version2.id.to_s, values
+  end
+
+  def test_fixed_version_filter_should_include_all_subproject_versions_when_filtering_any_subproject
+    with_settings :display_subprojects_issues => '0' do
+      version = Version.create!(:project => Project.find(3), :name => 'Any subproject version')
+      query = IssueQuery.new(:project => Project.find(1), :name => '_')
+      query.add_filter('subproject_id', '*', [''])
+      filter = query.available_filters["fixed_version_id"]
+      assert_not_nil filter
+      assert_include version.id.to_s, filter[:values].map(&:second)
+    end
+  end
+
   def test_query_with_multiple_custom_fields
     query = IssueQuery.find(1)
     assert query.valid?
-- 
2.43.0

